clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/agent/ReconcileClassicLinkSecurityGroupsAgent.java

```diff
@@ -0,0 +1,229 @@
+    this.amazonClientProvider = amazonClientProvider;
+    this.account = account;
+    this.region = region;
+    this.deployDefaults = deployDefaults;
+    this.pollIntervalMillis = pollIntervalMillis;
+    this.timeoutMillis = timeoutMillis;
+  }
+
+  @Override
+  public void run() {
+    if (!deployDefaults.isReconcileClassicLinkAccount(account)) {
+      return;
+    }
+    log.info("Checking classic link security groups in {}/{}", account.getName(), region);
+    AmazonEC2 ec2 = amazonClientProvider.getAmazonEC2(account, region, true);
+    List<String> classicLinkVpcIds = ec2.describeVpcClassicLink().getVpcs().stream().filter(VpcClassicLink::getClassicLinkEnabled).map(VpcClassicLink::getVpcId).collect(Collectors.toList());
+    if (classicLinkVpcIds.size() > 1) {
+      log.warn("Multiple classicLinkVpcs found: {}", classicLinkVpcIds);
+      throw new IllegalStateException("More than 1 classicLinkVpc found: " + classicLinkVpcIds);
+    }
+
+    if (classicLinkVpcIds.isEmpty()) {
+      return;
+    }
+    String classicLinkVpcId = classicLinkVpcIds.get(0);
+
+    RateLimiter apiRequestRateLimit = RateLimiter.create(5);
+    Collection<ClassicLinkInstance> classicLinkInstances = new LinkedList<>();
+    DescribeClassicLinkInstancesRequest request = new DescribeClassicLinkInstancesRequest().withMaxResults(1000);
+    while (true) {
+      apiRequestRateLimit.acquire();
+      DescribeClassicLinkInstancesResult result = ec2.describeClassicLinkInstances(request);
+      classicLinkInstances.addAll(result.getInstances());
+      if (result.getNextToken() == null) {
+        break;
+      }
+      request.setNextToken(result.getNextToken());
+    }
+
+    log.info("{} classic instances in {}/{}", classicLinkInstances.size(), account.getName(), region);
+
+    Map<String, String> groupNamesToIds = ec2.describeSecurityGroups(
+      new DescribeSecurityGroupsRequest()
+        .withFilters(
+          new Filter("vpc-id").withValues(classicLinkVpcId)))
+      .getSecurityGroups()
+      .stream()
+      .collect(Collectors.toMap(
+        SecurityGroup::getGroupName,
+        SecurityGroup::getGroupId));
+
+    reconcileInstances(ec2, groupNamesToIds, classicLinkInstances);
+  }
+
+  void reconcileInstances(AmazonEC2 ec2, Map<String, String> groupNamesToIds, Collection<ClassicLinkInstance> instances) {
+    RateLimiter apiRequestRateLimit = RateLimiter.create(5);
+    Map<String, String> groupIdsToNames = HashBiMap.create(groupNamesToIds).inverse();
+    StringBuilder report = new StringBuilder();
+    instances
+      .forEach(i -> {
+        List<String> existingClassicLinkGroups = i.getGroups().stream()
+          .map(GroupIdentifier::getGroupId)
+          .collect(Collectors.toList());
+
+        int maxNewGroups = deployDefaults.getMaxClassicLinkSecurityGroups() - existingClassicLinkGroups.size();
+        i.getTags()
+          .stream()
+          .filter(t -> AUTOSCALING_TAG.equals(t.getKey()))
+          .map(Tag::getValue)
```

Comment:
Might be a little difficult to follow what's going on here ... maybe a quick example comment for each map:

```
.map(Tag::getValue)  // clouddriver-main-v001
...
.map(this::getSecurityGroupNames) // clouddriver
```